language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -345,11 +364,18 @@ typeCheckM x term vTy = do
     vTermTy <- typeOf term
     when (vTermTy /= vTy) $ throwError (TypeMismatch x (void term) (getNormalizedType vTermTy) vTy)
 
--- | @unfoldFixOf vPat = NORM (vPat (fix vPat))@
+-- | @unfoldFixOf pat arg k = NORM (vPat (\(a :: k) -> ifix vPat a) arg)@
 unfoldFixOf
-    :: TyNameWithKind ()
-    -> NormalizedType TyNameWithKind ()
+    :: NormalizedType TyNameWithKind ()  -- ^ @vPat@
+    -> NormalizedType TyNameWithKind ()  -- ^ @vArg@
+    -> Kind ()                           -- ^ @k@
     -> TypeCheckM a (NormalizedType TyNameWithKind ())
-unfoldFixOf name vPat = do
-    let pat = getNormalizedType vPat
-    substituteNormalizeTypeTcm (TyFix () name <$> vPat) name pat
+unfoldFixOf pat arg k = do
+    let vPat = getNormalizedType pat
+        vArg = getNormalizedType arg
+    a <- liftQuote $ TyNameWithKind <$> freshTyName ((), k) "a"
+    normalizeTypeTcm $
+        foldl' (TyApp ()) vPat
+            [ TyLam () a k . TyIFix () vPat $ TyVar () a
```

Comment:
@michaelpj, that's the issue I mentioned in slack.

Previously it was

```haskell
substituteNormalizeTypeTcm (TyFix () name <$> vPat) name pat
```

and now we use

```haskell
foldl' (TyApp ()) vPat
```

here. The issue is that you pay gas in the new version while you did not pay gas in the old version. This is due to `TyIFix` not being a binder while `TyFix` was such. I believe this is a very minor thing considering how much pain and silliness it saves me not to deal with `ifix` being a binder, but this is still a semantic change, so we perhaps need to discuss is with the rest of the team. What do you think?

Reply:
> I'm confused about why you say you pay more gas - wouldn't we have done a substitution anyway, which would have revealed more normalization steps?

Yes, but it wouldn't be a "top-level" substitution. We pay for substitutions that happen inside `normalizeType`, but not for the top-level ones. This is because we pay not in `substituteNormalizeTypeM`, but in the definition of `normalizeTypeM` just before calling `substituteNormalizeTypeM`:

```haskell
normalizeTypeM (TyApp ann fun arg)           = do
    vFun <- normalizeTypeM fun
    vArg <- normalizeTypeM arg
    case getNormalizedType vFun of
        TyLam _ nArg _ body -> do
            countTypeNormalizationStep
            substituteNormalizeTypeM vArg nArg body
```

I'd prefer then to always pay in `substituteNormalizeTypeM`.

Reply:
That seems surprising to me... I guess I still don't quite understand our gas policy for typechecking. I don't see why we shouldn't pay for "top-level" substitutions, that sounds like a mistake.

Reply:
> That seems surprising to me... I guess I still don't quite understand our gas policy for typechecking. I don't see why we shouldn't pay for "top-level" substitutions, that sounds like a mistake.

I do no have any particular opinion here, but it's unrelated to this PR, so let's bring this problem to the public and decide later.